gossipd/test/run-onion_message.c

```diff
@@ -0,0 +1,580 @@
+					      om->reply_path->path[i]->encrypted_recipient_data));
+		}
+		printf("]}\n");
+	}
+	if (om->invoice)
+		json_strfield("invoice", tal_hex(tmpctx, om->invoice));
+	if (om->invoice_request)
+		json_strfield("invoice_request",
+			      tal_hex(tmpctx, om->invoice_request));
+	if (om->invoice_error)
+		json_strfield("invoice_error",
+			      tal_hex(tmpctx, om->invoice_error));
+}
+
+/* Return next onion (and updates blinding), or NULL */
+static u8 *json_test(const char *testname,
+		     const u8 *data,
+		     const struct privkey *me,
+		     const struct privkey *blinding_priv,
+		     struct pubkey *blinding)
+{
+	struct pubkey my_id, next_node;
+	struct secret ss, onion_ss;
+	struct pubkey ephemeral;
+	struct route_step *rs;
+	const u8 *cursor;
+	size_t max, maxlen;
+	struct onionpacket *op;
+	struct tlv_obs2_onionmsg_payload *om;
+
+	op = parse_onionpacket(tmpctx, data, tal_bytelen(data), NULL);
+	assert(op);
+
+	pubkey_from_privkey(me, &my_id);
+	printf("{");
+	json_strfield("test name", testname);
+	json_strfield("reader_privkey",
+		      type_to_string(tmpctx, struct privkey, me));
+	json_strfield("reader_id",
+		      type_to_string(tmpctx, struct pubkey, &my_id));
+
+	if (blinding_priv)
+		json_strfield("blinding_privkey",
+			      type_to_string(tmpctx, struct privkey,
+					     blinding_priv));
+	json_strfield("blinding",
+		      type_to_string(tmpctx, struct pubkey, blinding));
+	printf("\"onionmsg\": {\n");
+	json_strfield("raw", tal_hex(tmpctx, data));
+	json_strfield("version", tal_fmt(tmpctx, "%i", op->version));
+	json_strfield("public_key",
+		      type_to_string(tmpctx, struct pubkey, &op->ephemeralkey));
+	json_strfield("hop_payloads",
+		      tal_hex(tmpctx, op->routinginfo));
+	json_strfield("hmac",
+		      tal_hexstr(tmpctx, &op->hmac, sizeof(op->hmac)));
+	printf("},\n");
+
+	ephemeral = op->ephemeralkey;
+
+	/* Set this for test_ecdh */
+	mykey = me;
+	assert(unblind_onion(blinding, test_ecdh, &ephemeral, &ss));
+	json_strfield("ECDH shared secret",
+		      type_to_string(tmpctx, struct secret, &ss));
+	/* Reproduce internal calc from unblind_onion */
+	{
+		struct secret hmac;
+		subkey_from_hmac("blinded_node_id", &ss, &hmac);
+		json_strfield("HMAC256(\\\"blinded_node_id\\\", ss(i)) * k(i)",
+			      type_to_string(tmpctx, struct secret, &hmac));
+	}
+	json_strfield("Tweaked onion pubkey",
+		      type_to_string(tmpctx, struct pubkey, &ephemeral));
+
+	/* Now get onion shared secret and parse it. */
+	test_ecdh(&ephemeral, &onion_ss);
+	json_strfield("onion shared secret",
+		      type_to_string(tmpctx, struct secret, &onion_ss));
+	rs = process_onionpacket(tmpctx, op, &onion_ss, NULL, 0, false);
+	assert(rs);
+
+	printf("\"onion contents\": {\n");
+	json_strfield("raw", tal_hex(tmpctx, rs->raw_payload));
+
+	cursor = rs->raw_payload;
+	max = tal_bytelen(rs->raw_payload);
+	maxlen = fromwire_bigsize(&cursor, &max);
+	json_strfield("length", tal_fmt(tmpctx, "%"PRIu64, maxlen));
```

Comment:
Not sure that it can be an optimal solution
```suggestion
	json_strfield("length", tal_fmt(tmpctx, "%lu", maxlen));
```

Reply:
%zu is the correct format for size_t types, actually...